TST: add BLAS ILP64 run in Travis & Azure

pv opened the pull request:
Add Travis CI run for OpenBLAS ILP64 on Linux, and Windows & MacOS runs on Azure.

For Windows & Linux, I added the ILP64 build flags to an existing jobs. As there was only one MacOS job, I added a new job for that platform.

<s>The download location is currently temporary, possibly could make sense to move it to rackspace.</s> The openblas64_ builds are now on rackspace, together with the usual openblas ones.

pv committed:
CI: add openblas64_ downloading to openblas_support.py

pv committed:
CI: add travis openblas64_ target

tylerjereddy commented:
The new CI entry looks sensible at first glance based on the distutils output. We only have one specific test for the 64-bit integer support right now, right? And it has physical memory (and "slowness") requirements such that it may not be possible/desirable to run on a CI node anyway, right?

pv commented:
All BLAS/LAPACK calls use the 64-bit BLAS, so that the specific large memory test doesn't run is maybe not so bad. The CHECK_BLAS stuff will also check the 64bit blas gets linked in. Moreover this checks the build infra works.


On December 15, 2019 3:45:42 PM UTC, Tyler Reddy <notifications@github.com> wrote:
>The new CI entry looks sensible at first glance based on the distutils
>output. We only have one specific test for the 64-bit integer support
>right now, right? And it has physical memory (and "slowness")
>requirements such that it may not be possible/desirable to run on a CI
>node anyway, right?

charris commented:
Note that vdot (`vdot.c`) is currently chunked and calls the now modified `cblas_*dotc_sub` functions with `int` count.

pv commented:
In current master it actually falls to non-blas call. That's fixed in the other PR.

On December 15, 2019 6:09:49 PM UTC, Charles Harris <notifications@github.com> wrote:
>Note that vdot (`vdot.c`) is currently chunked and calls the now
>modified `cblas_*dotc_sub` functions with `int` count.

pv committed:
TST: linalg: add ilp64 lapack low-memory smoketest

LAPACK lwork call does not require much memory, and can be used as a
smoketest to whether LAPACK really uses 64-bit integers.

pv committed:
CI: use openblas64_ on azure windows

mattip commented on tools/openblas_support.py:
```diff
@@ -20,6 +20,21 @@
+    else:
+        raise RuntimeError("No uploaded OpenBLAS64_ for this arch!")
+
+
```
How did you build the 64-bit artifacts? Would you be interested in making a PR to produce these as part of the regular openblas-libs CI run?

pv replied:
It's in: https://github.com/MacPython/openblas-libs/pull/8

tylerjereddy commented:
I've merged https://github.com/MacPython/openblas-libs/pull/8 upstream just now. Sorry for the delay---I wanted to test the MacOS 64-bit OpenBLAS integer support locally using Pauli's new binaries + NumPy master, and they do indeed work just fine it seems. See notes there for how I tested.

Anyway, that should mean that the usual rackspace site binaries should be updated to include the 64-bit symbol-suffixed OpenBLAS within a few hours, after the CI runs on `openblas-libs` master, I think.

pv committed:
CI: add to Azure a MacOS ilp64 job

pv committed:
CI: use only one flag on travis for downloading openblas

pv commented:
Ok, this should be ready now. Also added macos.

mattip commented on tools/travis-before-install.sh:
```diff
@@ -5,12 +5,17 @@ free -m
+  else
+      sudo cp -r $target/64/lib/* /usr/lib
+      sudo cp $target/64/include/* /usr/include
+  fi
```
This seems awkward. Could we add an argument to `openblas_support.py:__main__` to  print another level in the directory, maybe `--get-root` or so?

pv replied:
Well, the openblas archive contents are a bit of a mess.
Made it to strip the common prefix directory instead.

mattip reviewed (commented):
A few nits, but looks good to me.

pv committed:
MAINT: make openblas_support.py strip path prefix from archive contents

Adapt CI scripts to this.

pv commented:
The pypy3 failure is unrelated: it's because today's pypy3 nightly build http://buildbot.pypy.org/nightly/py3.6/pypy-c-jit-98329-574325fb0c26-linux64.tar.bz2 does not contain a `lib` directory that ships the support libraries, whereas previous nightly builds (e.g. http://buildbot.pypy.org/nightly/py3.6/pypy-c-jit-98316-010bdf9668e6-linux64.tar.bz2) do contain it.

mattip commented:
The pypy download was invalid, it has now been fixed. There is a conflict with shippable.yml

pv commented:
Rebased.

pv commented:
Not sure what's going on with ppc64le on travis --- there were no changes to .travis.yml in the rebase but now it's complaining "An error occurred while generating the build script.". Maybe spurious failure.

mattip commented:
I think this PR is solid, but I am wondering how we will handle OpenBLAS library names going forward. See MacPython/gfortran-install#2. In short, there are 4 ways now to construct the name of the OpenBLAS library:
- MacPython/openblas-libs build the library and uses [`get_distutils_platform` and shell scripting](https://github.com/MacPython/openblas-libs/blob/master/travis-ci/build_steps.sh#L110) to create the name to be uploaded to rackspace
- CI for NumPy uses the internal script that is modified in this PR to get the name for downloading when running CI
- CI for SciPy [hard-codes](https://github.com/scipy/scipy/blob/master/.travis.yml#L149) the library names for macOS and doesn't seem to use the rackspace OpenBLAS on linux at all (that doesn't seem right, maybe I just don't see it)
- wheel building for both NumPy and SciPy uses the name as retrieved by [`get_gf_lib`](https://github.com/MacPython/numpy-wheels/blob/master/config.sh#L24), which is defined [here](https://github.com/MacPython/gfortran-install/blob/e9764d0a8f4e3adcd385c319e7742177193484c7/gfortran_utils.sh#L83) and uses the same `get_distutils_platform` as used when building the library, but without the extra shell scripting

Should we modify the library building, SciPy CI, and wheel building to use the `tools/openblas_support.py` file here? Or should we synchronize [`get_distutils_platform`](https://github.com/MacPython/gfortran-install/blob/e9764d0a8f4e3adcd385c319e7742177193484c7/gfortran_utils.sh#L8) for the `_64` prefix like in this PR and then use it in `openblas-libs` to get the name to upload?

pv commented:
The only "moving part" here is the platform/compiler tag, right? Perhaps it would be simplest if multibuild has a script (or function) that prints the whole platform tag suffix, that can then be used to e.g. set an environment variable?

I'd also consider the `64_` suffix as a part of the library name, ie. `openblas64_` is to `openblas` as `pybind11` is to `openblas`. So numpy probably should continue to have an openblas download script (there's numpy-specific stuff there anyway). The library downloading part that takes the library basename as argument, appends the platform tag and rackspace url, and unpacks somewhere, could perhaps be factored out to a separate script.

My main concern here is that it would be best to keep it as simple as possible, to make it easier to understand.

But anyway, I guess this is a bit orthogonal to this PR (which just continues status quo).

mattip commented:
Thanks for the thoughts. Agreed it is all orthagonal to this PR.

charris commented:
Let's give this a shot. Thanks Pauli.
